### PR TITLE
Added support for sending additional data in the ParticleEffect packet

### PR DESCRIPTION
**Pull request by worktycho**
Also the start of my refactor to move Broadcasts out of cWorld.

### CODE REVIEW

File: src/Bindings/LuaState.cpp [hunk 1]
```diff
@@ -1415,6 +1432,27 @@ bool cLuaState::CheckParamEnd(int a_Param)
...
+{
+	ASSERT(IsValid());
+	
+	tolua_Error tolua_err;
+	return tolua_isusertype(m_LuaState, a_Param, a_UserType.c_str(), 0, &tolua_err);
+}
+
+
+
+bool cLuaState::IsParamNumber(int a_Param)
```
Comment (madmaxoft) [inline src/Bindings/LuaState.cpp]: 5 empty lines is the standard. Not 3, not 6, but 5 :)

File: src/Bindings/LuaState.h [hunk 2]
```diff
@@ -259,6 +260,7 @@ class cLuaState
 	void GetStackValue(int a_StackPos, pRoot & a_Value);
 	void GetStackValue(int a_StackPos, pScoreboard & a_Value);
 	void GetStackValue(int a_StackPos, pWorld & a_Value);
+	void GetStackValue(int a_StackPos, pClientHandle & a_Value);
```
Comment (madmaxoft) [inline src/Bindings/LuaState.h]: Alpha-sort?

File: src/Broadcaster.cpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+{
+}
+
+
+void cBroadcaster::BroadcastParticleEffect(const AString & a_ParticleName, const Vector3f a_Src, const Vector3f a_Offset, float a_ParticleData, int a_ParticleAmount, cClientHandle * a_Exclude)
+{
+	m_World->DoWithChunkAt(a_Src,
+		[=](cChunk& a_Chunk) -> bool
+		{
+			for (auto&& client : a_Chunk.GetAllClients())
```
Comment (madmaxoft) [inline src/Broadcaster.cpp]: `for (auto && client: a_Chunk.GetAllClients())`

File: src/Broadcaster.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+cBroadcaster::cBroadcaster(cWorld * a_World) :
+	m_World(a_World)
+{
+}
+
+
+void cBroadcaster::BroadcastParticleEffect(const AString & a_ParticleName, const Vector3f a_Src, const Vector3f a_Offset, float a_ParticleData, int a_ParticleAmount, cClientHandle * a_Exclude)
+{
+	m_World->DoWithChunkAt(a_Src,
+		[=](cChunk& a_Chunk) -> bool
```
Comment (madmaxoft) [inline src/Broadcaster.cpp]: space

File: src/Broadcaster.cpp [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+			};
+			return true;
+		});
+}
+
+
+void cBroadcaster::BroadcastParticleEffect(const AString & a_ParticleName, const Vector3f a_Src, const Vector3f a_Offset, float a_ParticleData, int a_ParticleAmount, std::array<int, 2> a_Data, cClientHandle * a_Exclude)
+{
+	m_World->DoWithChunkAt(a_Src,
+		[=](cChunk& a_Chunk) -> bool
```
Comment (madmaxoft) [inline src/Broadcaster.cpp]: space

File: src/ChunkMap.cpp [hunk 6]
```diff
@@ -791,6 +791,28 @@ bool cChunkMap::DoWithChunk(int a_ChunkX, int a_ChunkZ, cChunkCallback & a_Callb
 }
 
 
+bool cChunkMap::DoWithChunkAt(Vector3i a_Pos, std::function<bool(cChunk&)> a_Callback)
```
Comment (madmaxoft) [inline src/ChunkMap.cpp]: Space before `&`

Comment (madmaxoft) [inline src/ChunkMap.cpp]: Also, is `a_Pos` in block coords or chunk coords? If in block coords, we customarily name it `a_BlockPos`.

File: src/ChunkMap.h [hunk 7]
```diff
@@ -104,6 +104,8 @@ class cChunkMap
 	/** Calls the callback for the chunk specified, with ChunkMapCS locked; returns false if the chunk doesn't exist, otherwise returns the same value as the callback */
 	bool DoWithChunk(int a_ChunkX, int a_ChunkZ, cChunkCallback & a_Callback);
 
+	bool DoWithChunkAt(Vector3i a_Pos, std::function<bool(cChunk&)> a_Callback);
```
Comment (madmaxoft) [inline src/ChunkMap.h]: Doxy-comment

File: src/Protocol/Protocol18x.cpp [hunk 8]
```diff
@@ -799,6 +799,44 @@ void cProtocol180::SendParticleEffect(const AString & a_ParticleName, float a_Sr
...
+		// blockcrack
+		// blockdust
+		case 37:
+		case 38:
+		{
+			Pkt.WriteVarInt32(static_cast<UInt32>(a_Data[0]));
+			break;
+		}
+		default:
+			break;
```
Comment (madmaxoft) [inline src/Protocol/Protocol18x.cpp]: Braces

File: src/Broadcaster.h [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+class cBroadcaster
+{
+
+public:
+
+	cBroadcaster(cWorld * a_World);
+
+	void BroadcastParticleEffect(const AString & a_ParticleName, const Vector3f a_Src, const Vector3f a_Offset, float a_ParticleData, int a_ParticleAmount, cClientHandle * a_Exclude = nullptr);
+
+	void BroadcastParticleEffect(const AString & a_ParticleName, const Vector3f a_Src, const Vector3f a_Offset, float a_ParticleData, int a_ParticleAmount, std::array<int, 2> a_Data, cClientHandle * a_Exclude = nullptr);
```
Comment (madmaxoft) [inline src/Broadcaster.h]: Is it wise, passing the vectors by-value?

Comment (worktycho) [inline src/Broadcaster.h]: Well, the previous api passed the coordinates as separate parameters which were just as large overall.
If you find any performance evidence though I'll change it.

**Comment by worktycho** [comment]
I've fixed the style issues.

File: src/Broadcaster.h (hunk 9 above)
Comment (madmaxoft) [inline src/Broadcaster.h]: True, that. It's most likely faster to copy 3 floats than to access them later via an indirection.

**Comment by madmaxoft** [comment]
Not yet, but close.

I like these changes, the cBroadcaster is a nice idea.

When you get this working, could you squash it all into a single commit?

**Comment by worktycho** [comment]
Those should be fixed now. I'll squash if CI passes.

**Comment by worktycho** [comment]
Squashed.

**Comment by madmaxoft** [comment]
You might want to link the "email address" you use for commits to your account on GitHub

**Pull request merged**